clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -513,10 +518,25 @@ ProducerIdAndEpoch producerIdAndEpoch() {
     }
 
     boolean producerIdOrEpochNotMatch(ProducerBatch batch) {
-        ProducerIdAndEpoch idAndEpoch = this.producerIdAndEpoch;
+        // Verifies that the producer id/epoch of the batch matches the current one
+        // of the partition.
+        ProducerIdAndEpoch idAndEpoch = topicPartitionBookkeeper.getPartition(batch.topicPartition).producerIdAndEpoch;
         return idAndEpoch.producerId != batch.producerId() || idAndEpoch.epoch != batch.producerEpoch();
     }
 
+    synchronized public void maybeUpdateProducerIdAndEpoch(TopicPartition topicPartition) {
+        if (!isTransactional())
+            topicPartitionBookkeeper.addPartition(topicPartition);
+
+        if (hasStaleProducerIdAndEpoch(topicPartition) && !hasInflightBatches(topicPartition)) {
```

Comment:
One option I was discussing with @bob-barrett is whether to continue using the old epoch. The benefit is that we don't need to drain all partitions because of a sequence error on a single partition. I'm not sure if there are any major downsides other than possibly making debugging more confusing. I guess for the transactional use case, it might be better to stick with the latest epoch when possible.

Reply:
Interesting point. My understanding is that we can't do this for the transactional use case so it might be better to stick with the latest epoch whenever possible as you said. However with [KIP-679](https://cwiki.apache.org/confluence/display/KAFKA/KIP-679%3A+Producer+will+enable+the+strongest+delivery+guarantee+by+default) on the way, this may be a nice optimization for the idempotent producer as, I suppose, an out of sequence for one partition would not affect the latency/throughput of the other partitions at all. If out of sequence errors are frequent, this may be worth. Do we believe that they are?

I suggest to tackle this in a separate PR anyway and keep this PR focused on fixing the identified issue.

Reply:
Yeah, I agree that this can be an optimization in a future PR